app/src/lib/git/stash.ts

```diff
@@ -68,17 +84,82 @@ export async function getDesktopStashEntries(
   return stashEntries
 }
 
+/**
+ * Returns the last Desktop created stash entry for the given branch
+ */
+export async function getLastDesktopStashEntryForBranch(
+  repository: Repository,
+  branchName: string
+) {
+  const entries = await getDesktopStashEntries(repository)
+
+  // Since stash objects are returned in a LIFO manner, the first
+  // entry found is guaranteed to be the last entry created
+  return entries.find(stash => stash.branchName === branchName) || null
+}
+
+/** Creates a stash entry message that idicates the entry was created by Desktop */
+export function createDesktopStashMessage(branchName: string, tipSha: string) {
+  return `${DesktopStashEntryMarker}<${branchName}@${tipSha}>`
+}
+
+/**
+ * Stash the working directory changes for the current branch
+ */
+export async function createDesktopStashEntry(
+  repository: Repository,
+  branchName: string,
+  tipSha: string
+) {
+  const message = createDesktopStashMessage(branchName, tipSha)
+  await git(
+    ['stash', 'push', '-m', message],
+    repository.path,
+    'createStashEntry'
+  )
+}
+
+/**
+ * Removes the stash entry identified by `stashSha`
+ */
+export async function dropDesktopStashEntry(
+  repository: Repository,
+  stashSha: string
+) {
+  if (stashSha === '') {
+    // the drop sub-command behaves like pop when no stash is given
+    // so we return if given an empty string to avoid that
+    return
+  }
+
+  const entry = ['stash@{', stashSha, '}'].join('')
+  await git(['stash', 'drop', entry], repository.path, 'dropStashEntry')
+}
+
+/**
+ * Applies the stash entry identified by matching `stashSha` to its commit hash.
+ *
+ * To see the commit hash of stash entry, run
+ * `git log -g refs/stash --pretty="%nentry: %gd%nsubject: %gs%nhash: %H%n"`
+ * in a repo with some stash entries.
+ */
+export async function applyStashEntry(
+  repository: Repository,
+  stashSha: string
+): Promise<void> {
+  await git(
```

Comment:
Are there any error cases here that you care about when applying the stash? Or is this covered by future work?

Reply:
I added a base test case in ac67324 and I plan to handle other errors in a later PR. Many of the error cases I'd have to worry about will be taken care of in higher levels of the app by requiring certain invariants be met before calling these _risky_ git functions.